File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        HashMap<Integer, Integer> sumCounts = new HashMap<>();
        sumCounts.put(0, 1);
        int currSum = 0, count = 0;
        for(int i = 0; i < nums.length; i++) {
            currSum += nums[i];
            int targetPrefixSum = currSum - k;
            if(sumCounts.containsKey(targetPrefixSum)) {
                count += sumCounts.get(targetPrefixSum);
            }
            sumCounts.put(currSum, sumCounts.getOrDefault(currSum, 0) + 1);
        }
        return count;
    }
}

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        count = 0
        curr_sum = 0
        prefix_counts = {0: 1}
        for num in nums:
            curr_sum += num
            if curr_sum - k in prefix_counts:
                count += prefix_counts[curr_sum - k]
            prefix_counts[curr_sum] = prefix_counts.get(curr_sum, 0) + 1

        return count